### PR TITLE
Issue 781

### PR DESCRIPTION
**Pull request by johannesherschel**
The constraint dialog now provides proposals for features with spaces and features equal to operator names when typing a quotation mark. Proposals are now also dependent on previous text, i.e. operators are proposed after a feature and vice versa.

### CODE REVIEW

**Comment by tthuem** [comment]
> Proposals are now also dependent on previous text, i.e. operators are proposed after a feature and vice versa.

We have had this functionality already implemented before. The content assist has recognized what is expected next. Why is there a new implemented of it?

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/ConstraintContentProposalProvider.java [hunk 1]
```diff
@@ -94,108 +92,134 @@ public ConstraintContentProposalProvider(Collection<String> featureNames) {
...
-		// Hide features for "A |"
-		for (final String s : featureList) {
-			proposals.add(new ContentProposal(s + (operatorNamesInFeatures.contains(s.trim()) ? " " + Features.FEATURE_SUFFIX : "")));
+		if (context.featureBefore) {
+			if (!context.quotationMark) {
+				// Add binary operators only iff their appearance makes sense in content proposal
+				// Example:
+				// Show "and" for "A |"
+				// Hide "and" for "A and |"
+				proposals.add(new ContentProposal("and"));
```
Comment (SundermannC) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/ConstraintContentProposalProvider.java:165]: Inconsistent usage of constants. Please also use constant strings for "and" and "or"

**Comment by Rahel12342** [comment]
It was not already implemented. Theoretically yes, practically there was a method supposed to do so (getProposalList()) which had a lot of TODOs and just returned a list with all proposals instead of filtered ones. See the changes we made in getProposalList() in ConstraintContentProposalProvider.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/ConstraintContentProposalProvider.java [hunk 2]
```diff
@@ -94,108 +92,134 @@ public ConstraintContentProposalProvider(Collection<String> featureNames) {
...
 		}
 
 		return proposals;
 	}
+
+	/**
+	 *
+	 * Summarizes necessary information about the current context of proposals, i.e. the currently typed word and the word before.
+	 *
+	 * @author Johannes Herschel
```
Comment (SundermannC) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/ConstraintContentProposalProvider.java:202]: Only one author. intended?

**Review by SundermannC — CHANGES_REQUESTED** [review]
Looks good, only one detail should be changed

**Comment by tthuem** [comment]
Please have a look at older versions. This has been implemented before. I would start at a three year old version, download and test it. Then go to newer or older versions depending on whether it works or not. Then, take a look at the code and inspect why it was removed.

**Comment by johannesherschel** [comment]
The functionality apparently was removed with de9b9e0 which noted that it was buggy, and 633664b added TODO notes on the desired behaviour. The old implementation indeed looks buggy to me.

**Comment by tthuem** [comment]
Thanks for double checking. @pinnecke did a heavy refactoring back then and apparently after six years we are still struggling to get everything working as it was before.

**Review by SundermannC — APPROVED** [review]

**Pull request merged**